src/lib/frontmatter.ts

```diff
@@ -0,0 +1,157 @@
+      error => error.instancePath === '/parent/uri' && error.keyword === 'pattern',
+    );
+
+    if (parentUriPatternErrorIndex >= 0) {
+      this.debug('removing category uri pattern error');
+      errors.splice(parentUriPatternErrorIndex, 1);
+    }
+  }
+
+  let changeCount = 0;
+  const unfixableErrors: ErrorObject[] = [];
+  const updatedData = structuredClone(data);
+
+  if (typeof errors === 'undefined' || !errors.length) {
+    return { changeCount, errors: [], hasIssues: false, unfixableErrors, updatedData };
+  }
+
+  errors.forEach(error => {
+    if (error.instancePath === '/category' && error.keyword === 'type') {
+      const uri = mappings.categories[data.category as string];
+      updatedData.category = {
+        uri: uri || `uri-that-does-not-map-to-${data.category}`,
+      };
+      changeCount += 1;
+    } else if (error.keyword === 'additionalProperties') {
+      const badKey = error.params.additionalProperty as string;
+      const extractedValue = data[badKey];
+      if (error.schemaPath === '#/additionalProperties') {
+        // if the bad property is at the root level, delete it
+        delete updatedData[badKey];
+        changeCount += 1;
+        if (badKey === 'excerpt') {
+          // if the `content` object exists, add to it. otherwise, create it
+          if (typeof updatedData.content === 'object' && updatedData.content) {
+            (updatedData.content as Record<string, unknown>).excerpt = extractedValue;
+          } else {
+            updatedData.content = {
+              excerpt: extractedValue,
+            };
+          }
+        } else if (badKey === 'categorySlug') {
+          updatedData.category = {
+            uri: extractedValue,
+          };
+        } else if (badKey === 'parentDoc') {
+          const uri = mappings.parentPages[extractedValue as string];
+          if (uri) {
+            updatedData.parent = {
+              uri,
+            };
+          }
+        } else if (badKey === 'parentDocSlug') {
+          updatedData.parent = {
+            uri: extractedValue,
+          };
+        } else if (badKey === 'hidden') {
+          updatedData.privacy = { view: extractedValue ? 'anyone_with_link' : 'public' };
+        } else if (badKey === 'order') {
+          updatedData.position = extractedValue;
+        }
+      } else {
+        unfixableErrors.push(error);
+      }
+    } else {
+      unfixableErrors.push(error);
+    }
+  });
+
+  return { errors, changeCount, hasIssues: true, unfixableErrors, updatedData };
+}
+
+export function writeFixes(
+  this: CommandClass['prototype'],
+  /** all metadata for the page that will be written to */
+  metadata: PageMetadata,
+  /** frontmatter changes to be applied */
+  updatedData: PageMetadata['data'],
+) {
+  this.debug(`writing fixes to ${metadata.filePath}`);
+  const result = grayMatter.stringify(metadata.content, updatedData);
+  fs.writeFileSync(metadata.filePath, result, { encoding: 'utf-8' });
```

Comment:
Wonder if we need to do some detection on these files to preserve their existing encoding. I know that `latin-1` doesn't support emojis, and saving an emoji into one results in some garbage text, wonder it converting that then to UTF-8 would fix that emoji or add more garbage.

Reply:
Ran a small test and we thankfully don't do any corruption on it so this should be fine.

```ts
const fs = require('node:fs');

const filePath = 'debug.md';
const content = `---
name: test
---
shrimp is bugs 🦐`;

fs.writeFileSync(filePath, content, { encoding: 'latin1' });

const file = fs.readFileSync(filePath, { encoding: 'utf-8' });
fs.writeFileSync(filePath, file, { encoding: 'utf-8' });
```

```
---
name: test
---
shrimp is bugs >�
```